File: Drawflow.Server/Models/Drawing.cs
namespace Drawflow.Server.Models;

using System.Text.Json.Serialization;

public class Drawing
{
    [JsonPropertyName("drawflow")]
    public DrawflowSection Drawflow { get; set; } = new();

    public class DrawflowSection
    {
        [JsonPropertyName("Home")]
        public DrawflowDataSection Home { get; set; } = new();

        [JsonPropertyName("Other")]
        public DrawflowDataSection Other { get; set; } = new();

        public class DrawflowDataSection
        {
            [JsonPropertyName("data")]
            public Dictionary<string, DrawflowDataElement> Data { get; set; } = new(StringComparer.OrdinalIgnoreCase);

            public class DrawflowDataElement
            {
                [JsonPropertyName("class")]
                public string? Class { get; set; }

                [JsonPropertyName("data")]
                public Dictionary<string, object> Data { get; set; } = new(StringComparer.OrdinalIgnoreCase);

                [JsonPropertyName("html")]
                public string? Html { get; set; }

                [JsonPropertyName("id")]
                public long Id { get; set; }

                [JsonPropertyName("inputs")]
                public Dictionary<string, DrawflowConnectionsElement> Inputs { get; set; } = new(StringComparer.OrdinalIgnoreCase);

                [JsonPropertyName("name")]
                public string? Name { get; set; }

                [JsonPropertyName("outputs")]
                public Dictionary<string, DrawflowConnectionsElement> Outputs { get; set; } = new(StringComparer.OrdinalIgnoreCase);

                [JsonPropertyName("pos_x")]
                public double PosX { get; set; }

                [JsonPropertyName("pos_y")]
                public double PosY { get; set; }

                [JsonPropertyName("typenode")]
                public bool TypeNode { get; set; }

                public class DrawflowConnectionsElement
                {
                    [JsonPropertyName("connections")]
                    public List<DrawflowConnectionElement> Connections { get; set; } = [];

                    public class DrawflowConnectionElement
                    {
                        [JsonPropertyName("input")]
                        public string? Input { get; set; }

                        [JsonPropertyName("node")]
                        public string? Node { get; set; }

                        [JsonPropertyName("output")]
                        public string? Output { get; set; }
                    }
                }
            }
        }
    }
}

File: drawflow.client/src/app/features/drawflow/models/drawing.ts
export interface Drawing {
  Home: DrawflowSection;
  Other: DrawflowSection;
}

export interface DrawflowSection {
  data: { [key: string]: DrawflowDataElement };
}

export interface DrawflowDataElement {
  id: number;
  name: string | undefined | null;
  data: object;
  class: string | undefined | null;
  html: string | undefined | null;
  typenode: boolean;
  inputs: { [key: string]: DrawflowConnectionsElement };
  outputs: { [key: string]: DrawflowConnectionsElement };
  pos_x: number;
  pos_y: number;
}

export interface DrawflowConnectionsElement {
  connections: DrawflowConnectionElement[];
}

export interface DrawflowConnectionElement {
  node: string;
  input: string | undefined | null;
  output: string | undefined | null;
}
